crates/sqllib/src/array.rs

```diff
@@ -643,3 +643,277 @@ where
+where
+    T: Eq + Clone + Hash + Ord,
+{
+    let left = left?;
+    Some(array_except__(left, right))
+}
+
+#[doc(hidden)]
+pub fn array_except_N<T>(left: Vec<T>, right: Option<Vec<T>>) -> Option<Vec<T>>
+where
+    T: Eq + Clone + Hash + Ord,
+{
+    let right = right?;
+    Some(array_except__(left, right))
+}
+
+#[doc(hidden)]
+pub fn array_exceptNN<T>(left: Option<Vec<T>>, right: Option<Vec<T>>) -> Option<Vec<T>>
+where
+    T: Eq + Clone + Hash + Ord,
+{
+    let left = left?;
+    let right = right?;
+    Some(array_except__(left, right))
+}
+
+#[doc(hidden)]
+pub fn array_union__<T>(left: Vec<T>, right: Vec<T>) -> Vec<T>
+where
+    T: Eq + Clone + Hash + Ord,
+{
+    let left = to_set(&left);
+    let right = to_set(&right);
+    let result = left.union(&right);
+    let mut result = result.cloned().collect::<Vec<T>>();
+    result.sort();
+    result
+}
+
+#[doc(hidden)]
+pub fn array_unionN_<T>(left: Option<Vec<T>>, right: Vec<T>) -> Option<Vec<T>>
+where
+    T: Eq + Clone + Hash + Ord,
+{
+    let left = left?;
+    Some(array_union__(left, right))
+}
+
+#[doc(hidden)]
+pub fn array_union_N<T>(left: Vec<T>, right: Option<Vec<T>>) -> Option<Vec<T>>
+where
+    T: Eq + Clone + Hash + Ord,
+{
+    let right = right?;
+    Some(array_union__(left, right))
+}
+
+#[doc(hidden)]
+pub fn array_unionNN<T>(left: Option<Vec<T>>, right: Option<Vec<T>>) -> Option<Vec<T>>
+where
+    T: Eq + Clone + Hash + Ord,
+{
+    let left = left?;
+    let right = right?;
+    Some(array_union__(left, right))
+}
+
+#[doc(hidden)]
+pub fn array_intersect__<T>(left: Vec<T>, right: Vec<T>) -> Vec<T>
+where
+    T: Eq + Clone + Hash + Ord,
+{
+    let left = to_set(&left);
+    let right = to_set(&right);
+    let result = left.intersection(&right);
+    let mut result = result.cloned().collect::<Vec<T>>();
+    result.sort();
+    result
+}
+
+#[doc(hidden)]
+pub fn array_intersectN_<T>(left: Option<Vec<T>>, right: Vec<T>) -> Option<Vec<T>>
+where
+    T: Eq + Clone + Hash + Ord,
+{
+    let left = left?;
+    Some(array_intersect__(left, right))
+}
+
+#[doc(hidden)]
+pub fn array_intersect_N<T>(left: Vec<T>, right: Option<Vec<T>>) -> Option<Vec<T>>
+where
+    T: Eq + Clone + Hash + Ord,
+{
+    let right = right?;
+    Some(array_intersect__(left, right))
+}
+
+#[doc(hidden)]
+pub fn array_intersectNN<T>(left: Option<Vec<T>>, right: Option<Vec<T>>) -> Option<Vec<T>>
+where
+    T: Eq + Clone + Hash + Ord,
+{
+    let left = left?;
+    let right = right?;
+    Some(array_intersect__(left, right))
+}
+
+// The result type must always be Vec<Option<T>>
+// The result type must always be Vec<Option<T>>
+#[doc(hidden)]
+pub fn array_insert____<T>(array: Vec<T>, pos: i32, value: T) -> Vec<Option<T>>
+where
+    T: Clone,
+{
+    array_insert___N(array, pos, Some(value))
+}
+
+#[doc(hidden)]
+#[allow(clippy::needless_range_loop)]
+pub fn array_insert___N<T>(array: Vec<T>, pos: i32, value: Option<T>) -> Vec<Option<T>>
```

Comment:
Here, if the array is say, an array of nullable i32, `Vec<Option<i32>>`, then:
`T actually is: Option<i32>`
`value: Option<Option<i32>>`
And the result: `Vec<Option<Option<i32>>>`

We need a function like:

```rust
fn array_insert_N__(array: Vec<Option<T>>, pos: i32, value: T>);
fn array_insert_N_N(array: Vec<Option<T>>, pos: i32, value: Option<T>);
```